PathCG should support continuous rounded rectangles

lilyspiniolas opened the pull request:
#### 52716b6f551e192a6aaa3d9e02b5be4513c51ed7
<pre>
PathCG should support continuous rounded rectangles
<a href="https://bugs.webkit.org/show_bug.cgi?id=288127">https://bugs.webkit.org/show_bug.cgi?id=288127</a>
<a href="https://rdar.apple.com/145233717">rdar://145233717</a>

Reviewed by Aditya Keerthi.

Added a new path segment type `PathContinuousRoundedRect`
and support for drawing them via CoreGraphics.
For other platform paths, the segments will be drawn
as normal rounded rects.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::addContinuousRoundedRect):
(WebCore::Path::singleContinuousRoundedRect const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleContinuousRoundedRect const):
* Source/WebCore/platform/graphics/PathSegment.h:
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::PathContinuousRoundedRect::calculateEndPoint const):
(WebCore::PathContinuousRoundedRect::tryGetEndPointWithoutContext const):
(WebCore::PathContinuousRoundedRect::extendFastBoundingRect const):
(WebCore::PathContinuousRoundedRect::extendBoundingRect const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::add):
(WebCore::PathStream::singleContinuousRoundedRect const):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::add):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::add):
(WebCore::addToCGContextPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::add):
* Source/WebCore/platform/graphics/skia/PathSkia.h:
* Source/WebKit/GPUProcess/graphics/PathSegment.serialization.in:

Canonical link: <a href="https://commits.webkit.org/290828@main">https://commits.webkit.org/290828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8cbfcde139ab760ddf2316d33448225a90941a2f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96021 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41789 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93044 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10932 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18852 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69963 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27483 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93993 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8313 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82460 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50289 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8086 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40913 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83817 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78386 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97994 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89769 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18195 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13366 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78967 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18455 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78289 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78164 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19359 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22653 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11454 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18202 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23545 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112339 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17938 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32636 "Found 5 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.default, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21394 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19723 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -310,6 +321,20 @@ static inline void addToCGContextPath(CGContextRef context, PathRoundedRect segm
+static inline void addToCGContextPath(CGContextRef context, PathContinuousRoundedRect segment)
+{
+    // No API to add continuous rounded rects to context.
+    auto path = PathCG::create();
```
Prefer explicit smart pointer:
```suggestion
    Ref path = PathCG::create();
```

pxlcoder commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -296,6 +297,16 @@ void PathCG::add(PathRoundedRect roundedRect)
+    return;
+#endif
+    // Continuous rounded rects are unavailable. Paint a normal rounded rect instead.
+    add(PathRoundedRect { continuousRoundedRect.rect, FloatSize { continuousRoundedRect.cornerWidth, continuousRoundedRect.cornerHeight } });
```
```suggestion
#else
    // Continuous rounded rects are unavailable. Paint a normal rounded rect instead.
    add(PathRoundedRect { continuousRoundedRect.rect, FloatSize { continuousRoundedRect.cornerWidth, continuousRoundedRect.cornerHeight } });
#endif
```

lilyspiniolas committed:
PathCG should support continuous rounded rectangles
https://bugs.webkit.org/show_bug.cgi?id=288127
rdar://145233717

Reviewed by Aditya Keerthi.

Added a new path segment type `PathContinuousRoundedRect`
and support for drawing them via CoreGraphics.
For other platform paths, the segments will be drawn
as normal rounded rects.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::addContinuousRoundedRect):
(WebCore::Path::singleContinuousRoundedRect const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleContinuousRoundedRect const):
* Source/WebCore/platform/graphics/PathSegment.h:
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::PathContinuousRoundedRect::calculateEndPoint const):
(WebCore::PathContinuousRoundedRect::tryGetEndPointWithoutContext const):
(WebCore::PathContinuousRoundedRect::extendFastBoundingRect const):
(WebCore::PathContinuousRoundedRect::extendBoundingRect const):
(WebCore::operator<<):
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::add):
(WebCore::PathStream::singleContinuousRoundedRect const):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::add):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::add):
(WebCore::addToCGContextPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::add):
* Source/WebCore/platform/graphics/skia/PathSkia.h:
* Source/WebKit/GPUProcess/graphics/PathSegment.serialization.in:

Canonical link: https://commits.webkit.org/290828@main
